README.md
# ESPOL - Fundamentos de Programación


## Script de validación de datos
Este script le servirá a cada profesor de Fundamentos de Programaciòn para validar el formato de los datos que se solicitan al finalizar el semestre. 

El script deberá variar para cada semestre según las ponderaciones de los temas de los examenes, proyectos y sustentaciones. 

**Datos a validar(2017-II):**

* Nombre
* Matricula
* Genero
* Paralelo
* cod_carrera
* veces_tomadas
* 1er_proyecto
* 1er_sustent
* 1er_lecciones
* 1er_calif\_final
* 1er_exam\_tema1
* 1er_exam\_tema2
* 1er_exam\_tema3
* 1er_exam\_tema4
* 2do_proyecto
* 2do_sustent
* 2do_lecciones
* 2do_calif\_final
* 2do_exam\_tema1
* 2do_exam\_tema2
* 2do_exam\_tema3
* calif_final\_practica
* 3er_proyecto
* 3er_calif\_final
* 3er_exam\_tema1
* 3er_exam\_tema2
* 3er_exam\_tema3

**Validaciones implementadas**

| Tipo de error    | Color    | Soluciona | Comentarios |
| ---------------  |----------| :------:| ----------------|
| Vacio            | Naranja  |         | Los datos básicos (Nombre, paralelo, ...) no deben estar vacios |
| Opción no valida | Azul     |         | Las celdas género y veces_tomadas deben estar entre los valores posibles |
| Fuera de rango   | Amarillo |         | La calificación debe estar dentro del rango posible |
| Tipo no numérico | Verde    |  X     | Todas las calificacoines deben ser entero o float |
| No redondeado    | Rosado   |  X     | Esto aplica para las calificaciones finales(El script lo corrige) |

## Dependencias

Install the next python packages:

* [Openpyxl](https://openpyxl.readthedocs.io/en/stable/)

## Uso

* Complete los datos de su paralelo de fundamentos en el archivo **PARNN_AÑOTERMINO.xlsx**
* Corra el script **validation_data.py** pero modifique el nombre del archivo en la línea 8. Al finalizar se imprimirá un mensaje con la cantidad de estudiantes en los que se encontraron errores.
* Abra su archivo y verifique si hay celdas con los colores descritos en la tabla descrita anteriormente.
* Corrija los errores y vuelva a correr el script.

======

© 2018 eslozano

CHANGES.md
Change Log
==========

2018-09-21: v1.4
--------------------

* Fixes column bug. The script select only data of N columns, not all anymore. 
* Reported by Jose Villa 


2018-09-06: v1.3
--------------------

* The script fixes two type of errors: Not rounded values and value not numeric.


2018-03-06: v1.2
--------------------

* Adding two new columns to excel file and script (First and Second Partial Lessons)


2018-03-02: v1.1
--------------------

* Creating git repo.
* Implementing data validation script first version :)

src/estudiante.py
import random as rd

class Estudiante:
	nombre = None
	matricula = None
	genero = None
	paralelo = None
	cod_carrera = None
	veces_tomadas= None
	sustent = None
	proyecto = None
	examen = None
	calif_final = None
	lecciones = None

	def __init__(self, datos):
		#La tupla contiene el objeto celda del excel para cada variable
		nombre,matricula,genero,paralelo,cod_carrera,veces_tomadas, \
			primer_proyecto,primer_sustent,primer_lecciones,primer_calif_final, \
			primer_exam_tema1,primer_exam_tema2,primer_exam_tema3, \
			segundo_proyecto,segundo_sustent,segundo_lecciones,segundo_calif_final, \
			segundo_exam_tema1,segundo_exam_tema2,segundo_exam_tema3, \
			calif_final_practica, \
			tercer_proyecto,tercer_calif_final, \
			tercer_exam_tema1,tercer_exam_tema2,tercer_exam_tema3 = datos
		self.nombre = nombre
		self.matricula = matricula
		self.genero = genero
		self.paralelo = paralelo
		self.cod_carrera = cod_carrera
		self.veces_tomadas = veces_tomadas
		self.actividades = {
			"sustent" : {
				"1er_sustent": primer_sustent, 
				"2do_sustent": segundo_sustent
			},
			"proyecto" : {
				"1er_proyecto":primer_proyecto,
				"2do_proyecto":segundo_proyecto,
				"3er_proyecto":tercer_proyecto
			},
			"lecciones" : { 
				"1er_lecciones":primer_lecciones,
				"2do_lecciones": segundo_lecciones
			},
			"calif_final":{
				"1er_calif_final":primer_calif_final,
				"2do_calif_final":segundo_calif_final,
				"3er_calif_final":tercer_calif_final,
				"calif_final_practica": calif_final_practica
			}		
		}

		self.examen = {
			"1er_exam_": { "tema1": primer_exam_tema1 , "tema2": primer_exam_tema2 , "tema3": primer_exam_tema3 },
			"2do_exam_": { "tema1": segundo_exam_tema1 , "tema2": segundo_exam_tema2 , "tema3": segundo_exam_tema3 },
			"3er_exam_": { "tema1": tercer_exam_tema1 , "tema2": tercer_exam_tema2 , "tema3": tercer_exam_tema3 }
		}


	def validarDatoVacio(self,dato,VACIO):
		if not dato and dato!=0:
			return VACIO

	def validarGenero(self,genero,OPCION_NO_VALIDA, generos):
		if genero not in generos:
			return OPCION_NO_VALIDA

	def validarVecesTomadas(self,veces_tomadas,OPCION_NO_VALIDA,veces):
		if veces_tomadas and veces_tomadas not in veces:
			return OPCION_NO_VALIDA

	def validarCalificacion(self,calificacion,FUERA_DE_RANGO, max):
		if not 0<=calificacion<=max:
			return FUERA_DE_RANGO

	def validarRedondeo(self,calificacion,NO_REDONDEADO):
		if isinstance(calificacion, float):
			return NO_REDONDEADO

	def validarNumero(self,calificacion,TIPO_NO_NUMERICO):
		if not isinstance(calificacion,float) and not isinstance(calificacion, int):
			return TIPO_NO_NUMERICO

	def validarDatos(self,errores_mensajes,validaciones):
		VACIO,OPCION_NO_VALIDA,FUERA_DE_RANGO,TIPO_NO_NUMERICO,NO_REDONDEADO = errores_mensajes #Editar en caso de agregar mas mensajes
		d = {}
		d['nombre'] = self.validarDatoVacio(self.nombre,VACIO)
		d['matricula'] = self.validarDatoVacio(self.matricula,VACIO)
		d['paralelo'] = self.validarDatoVacio(self.paralelo,VACIO)
		d['cod_carrera'] = self.validarDatoVacio(self.cod_carrera,VACIO)
		d['veces_tomadas'] = self.validarVecesTomadas(self.veces_tomadas,OPCION_NO_VALIDA,validaciones['veces_tomadas'])
		#Variables con más validaciones
		d['genero'] = self.validarDatoVacio(self.genero,VACIO) or self.validarGenero(self.genero,OPCION_NO_VALIDA, validaciones['genero'])

		for actividad, califD in self.actividades.items():
			for nombre,calif in califD.items():
				if calif:
					if actividad=="proyecto":
						valorValidacion = validaciones['proyecto'][nombre]
					else:
						valorValidacion = validaciones[actividad]
					validacion = self.validarNumero(calif,TIPO_NO_NUMERICO) or self.validarCalificacion(calif, FUERA_DE_RANGO, valorValidacion )
					if actividad=="calif_final":
						validacion = validacion or self.validarRedondeo(calif,NO_REDONDEADO)
					d[nombre] = validacion

		for examen, temas in self.examen.items():
			for tema, calif in temas.items():
				if calif:
					d[examen+tema] = self.validarNumero(calif,TIPO_NO_NUMERICO) or self.validarCalificacion(calif, FUERA_DE_RANGO, validaciones['examen'][examen][tema])
		return d

	def validar(self, errores_mensajes,validaciones):
		validaciones = self.validarDatos(errores_mensajes,validaciones)
		errores  = {}
		for e in errores_mensajes:
			errores[e]=[]
		for campo, validacion in validaciones.items():
			if validacion:
				errores[validacion].append(campo)
		return errores

	def convertir(self):
		for actividad, califD in self.actividades.items():
			for nombre,calif in califD.items():
				if calif:
					if actividad == "calif_final": #Notas finales siempre deben ser redondeadas
						valor = int(calif)
					else:
						valor = float(calif)
					self.actividades[actividad][nombre] = valor

		for examen, temas in self.examen.items():
			for tema, calificacion in temas.items():
				if calificacion:
					self.examen[examen][tema] = float(calificacion)

fundamentos_validate_data.py
from openpyxl import load_workbook
from openpyxl.styles import PatternFill
from estudiante import Estudiante
import pandas as pd

OPCION_NO_VALIDA = 'dato_no_valido'
VACIO = 'dato_vacio'
FUERA_DE_RANGO = 'dato_fuera_de_rango'
NO_REDONDEADO = 'dato_no_redondeado'
TIPO_NO_DECIMAL = 'dato_no_decimal'



def fillCell(cell,color,fill_type=None):
	fill = PatternFill(fgColor=color,fill_type=fill_type)
	cell.fill = fill

def validateDataSheet(filename,columnas,colores,errorsMessages,validaciones):
	header_rows = 2
	try:
		wb = load_workbook(filename)
		data_sheet = wb["DATA"]
		total_students = data_sheet.max_row - header_rows
		total_columns = data_sheet.max_column 

		print("Total estudiantes a revisar: "+str(total_students))
		total_errors = 0
		for row in data_sheet.iter_rows(min_row=header_rows+1, max_col=total_columns, max_row=data_sheet.max_row):
			data = []
			for cell in row:
				data.append(cell.value)
				fillCell(cell,colores[errorsMessages[0]]) #Restore cell color
			estudiante = Estudiante(data)
			errores = estudiante.validar(errorsMessages,validaciones)
			countErrorStudent = 0
			for error, campos in errores.items():
				color = colores[error]
				countErrorStudent += len(campos)
				for c in campos:
					if c in columnas:
						i = columnas.index(c)
						fillCell(row[i],color,'solid')
			if countErrorStudent:
				print('{:<9}{}'.format('FALLÓ',estudiante.nombre))
				total_errors += 1
			else:
				print('{:<9}{}'.format('OK',estudiante.nombre))
		print("RESULTADOS: Se detectaron errores en las calificaciones de ", total_errors," estudiantes.")
		print("Revise en el archivo las celdas resaltadas de colores:")
		print("Naranja(vacio), Azul(opcion_no_valida), Amarillo(fuera_de_rango) o Verde(necesita_mejoamiento).")
		wb.save(filename)
	except IOError as e:
		print("ERROR: No." , e.errno , e )
		if e.errno == 2:
			print("El archivo ",filename," no se encuentra en la carpeta del script.")
		if e.errno == 13:
			print("El archivo ",filename," se encuentra abierto.")

def dataAnalysis(filename):
	df = pd.read_excel("filename",skiprows =2)



#Escriba el nombre del archivo de datos
filename = "PAR7_2017II.xlsx"
#Las columnas deben estar en el orden del excel
columnas = ["nombre","matricula","genero","paralelo","cod_carrera","veces_tomadas",
			"1er_proyecto", "1er_sustent", "1er_calif_final",
			"1er_exam_tema1","1er_exam_tema2","1er_exam_tema3","1er_exam_tema4",
			"2do_proyecto", "2do_sustent", "2do_calif_final",
			"2do_exam_tema1","2do_exam_tema2","2do_exam_tema3",
			"calif_final_practica","AP_sin_mejoramiento",
			"3er_proyecto","3er_calif_final",
			"3er_exam_tema1","3er_exam_tema2","3er_exam_tema3"
]
#Los colores en los que se resaltarán las celdas según los errores
colors = {
	VACIO:"ffd8bf", #naranja 
	OPCION_NO_VALIDA: '6aa2fc', #azul 
	FUERA_DE_RANGO: 'fffa00', #amarillo
	NO_REDONDEADO: '7cff94', #verde
	TIPO_NO_DECIMAL: 'ff7ff4' #rosado
}



errorsMessages= [VACIO,OPCION_NO_VALIDA,FUERA_DE_RANGO,NO_REDONDEADO,TIPO_NO_DECIMAL]

validaciones = {	
	"genero": ["F","M"],
	"veces_tomadas":[2,3],
	"calif_final": 100,
	"sustent": 1,
	"proyecto": {"1er_proyecto":20,"2do_proyecto":20,"3er_proyecto":25},
	"examen":{
		"1er_exam_": { "tema1": 20, "tema2": 30, "tema3": 45, "tema4": 10 },
		"2do_exam_": { "tema1": 35, "tema2": 55, "tema3": 10 },
		"3er_exam_": { "tema1": 40, "tema2": 55, "tema3": 10 } #el tema 2 tiene bono de 5 puntos
	}
}

validateDataSheet(filename,columnas,colors,errorsMessages, validaciones)

src/validation_data.py
from openpyxl import load_workbook
from openpyxl.styles import PatternFill
from estudiante import Estudiante
import pandas as pd


#Escriba el nombre del archivo de los datos a validar
filename = "PARNN_2018I.xlsx"




OPCION_NO_VALIDA = 'dato_no_valido'
VACIO = 'dato_vacio'
FUERA_DE_RANGO = 'dato_fuera_de_rango'
NO_REDONDEADO = 'dato_no_redondeado'
TIPO_NO_NUMERICO = 'dato_no_numerico'



def fillCell(cell,color,fill_type=None):
	fill = PatternFill(fgColor=color,fill_type=fill_type)
	cell.fill = fill

def convertirDatosNumericos(estudiante,row,columnas):
	for i,columna in enumerate(columnas):
		if "exam" in columna:
			for examen, temas in estudiante.examen.items():
				for tema, calif in temas.items():
					if calif and examen+tema == columna: 
						row[i].value = calif
		else:
			for actividad,dataD in estudiante.actividades.items():
				for item,calif in dataD.items():
					if calif and item==columna:
						row[i].value = calif

def validateDataSheet(filename,columnas,colores,errorsMessages,validaciones):
	header_rows = 2
	VACIO = errorsMessages[0]
	NO_REDONDEADO = errorsMessages[-1]
	TIPO_NO_NUMERICO = errorsMessages[-2]
	try:
		wb = load_workbook(filename)
		data_sheet = wb["DATA"]
		total_students = data_sheet.max_row - header_rows
		total_columns = data_sheet.max_column 

		print("Total estudiantes a revisar: "+str(total_students))
		total_errors = 0
		for row in data_sheet.iter_rows(min_row=header_rows+1, max_col=total_columns, max_row=data_sheet.max_row):
			data = []
			for cell in row:
				data.append(cell.value)
				fillCell(cell,colores[VACIO]) #Restore cell color
			estudiante = Estudiante(data)
			estudiante.convertir()
			convertirDatosNumericos(estudiante,row,columnas)
			#Luego de convertir datos numéricos, se valida
			errores = estudiante.validar(errorsMessages,validaciones)
			countErrorStudent = 0
			for error, campos in errores.items():
				color = colores[error]
				countErrorStudent += len(campos)
				for columna in campos:
					if columna in columnas:
						i = columnas.index(columna)
						if error==NO_REDONDEADO: #El error detectado es corregido por el script
							row[i].value = estudiante.calif_final[columna]
						else:
							fillCell(row[i],color,'solid') #Errores no corregidos se pintan
			
			if countErrorStudent:
				print('{:<9}{}'.format('FALLÓ',estudiante.nombre))
				total_errors += 1
			else:
				print('{:<9}{}'.format('OK',estudiante.nombre))
		if total_errors:
			print("RESULTADOS: Se detectaron errores en las calificaciones de ", total_errors," estudiantes.")
			print()
			print("Revise su archivo de calificaciones, y corrija los errores de las CELDAS RESALTADAS con los siguientes colores:")
			print("1. Naranja \t --> \t Dato requerido.")
			print("2. Azul \t --> \t Valor no válido, revise la cabecera del archivo modelo para las posibles opciones.")
			print("3. Amarillo \t --> \t Valor fuera del rango posible.")
			#print("4. Verde \t --> \t El dato no es numérico.")
			print("="*150)
			print("**OJO: Los errores tipo dato_no_redondeado y dato_no_numerico fueron corregidos por el script")
		else:
			print("="*150)
			print('¡Felicidades! Tu archivo de excel está listo para ser enviado')
			print('Por favor enviar a rabonilla@espol.edu.ec y a eslozano@espol.edu.ec')
		print("="*150)
		print("Copyright (c) 2018 eslozano")
		print('All Rights Reserved :)')
		wb.save(filename)
	except IOError as e:
		print("ERROR: No." , e.errno , e )
		if e.errno == 2:
			print("El archivo ",filename," no se encuentra en la carpeta del script.")
		if e.errno == 13:
			print("El archivo ",filename," se encuentra abierto.")

def dataAnalysis(filename):
	df = pd.read_excel("filename",skiprows =2)


#Las columnas deben estar en el orden del excel
columnas = ["nombre","matricula","genero","paralelo","cod_carrera","veces_tomadas",
			"1er_proyecto", "1er_sustent","1er_lecciones", "1er_calif_final",
			"1er_exam_tema1","1er_exam_tema2","1er_exam_tema3",
			"2do_proyecto", "2do_sustent","2do_lecciones", "2do_calif_final",
			"2do_exam_tema1","2do_exam_tema2","2do_exam_tema3",
			"calif_final_practica",
			"3er_proyecto","3er_calif_final",
			"3er_exam_tema1","3er_exam_tema2","3er_exam_tema3"
]
#Los colores en los que se resaltarán las celdas según los errores
colors = {
	VACIO:"ffd8bf", #naranja 
	OPCION_NO_VALIDA: '6aa2fc', #azul 
	FUERA_DE_RANGO: 'fffa00', #amarillo
	TIPO_NO_NUMERICO: '96D701', #verde
	NO_REDONDEADO: 'ff7ff4' #rosado
}

errorsMessages= [VACIO,OPCION_NO_VALIDA,FUERA_DE_RANGO,TIPO_NO_NUMERICO,NO_REDONDEADO]

validaciones = {	
	"genero": ["F","M"],
	"veces_tomadas":[2,3],
	"calif_final": 100,
	"sustent": 1,
	"proyecto": {"1er_proyecto":20,"2do_proyecto":20,"3er_proyecto":25},
	"examen":{
		"1er_exam_": { "tema1": 40, "tema2": 50, "tema3": 10 },
		"2do_exam_": { "tema1": 40, "tema2": 50, "tema3": 10 },
		"3er_exam_": { "tema1": 40, "tema2": 50, "tema3": 10 }
	},
	"lecciones":10,
}

validateDataSheet(filename,columnas,colors,errorsMessages, validaciones)